Dependencies/xr/Source/ARCore/XR.cpp

```diff
@@ -642,6 +663,158 @@ namespace xr
         return m_impl->TryInitialize();
     }
 
+    arcana::background_dispatcher<64> scheduler;
+    arcana::task<bool, std::exception_ptr> System::IsSessionSupportedAsync(SessionType sessionType)
+    {
+        // Currently only AR is supported on Android
+        if (sessionType == SessionType::IMMERSIVE_AR)
+        {
+            // Spin up a background thread to own the polling check.
+            arcana::task_completion_source<bool, std::exception_ptr> tcs;
+            arcana::make_task(scheduler, arcana::cancellation::none(), [sessionType, tcs]() mutable
+            {
+                // Query ARCore to check if AR sessions are supported.
+                // If not yet installed then poll supported status up to 100 times over 20 seconds.
+                for (int i = 0; i < 100; i++)
+                {
+                    ArAvailability arAvailability{};
+                    ArCoreApk_checkAvailability(GetEnvForCurrentThread(), GetAppContext(), &arAvailability);
+                    switch (arAvailability)
+                    {
+                        case AR_AVAILABILITY_SUPPORTED_APK_TOO_OLD:
+                        case AR_AVAILABILITY_SUPPORTED_INSTALLED:
+                        case AR_AVAILABILITY_SUPPORTED_NOT_INSTALLED:
+                            tcs.complete(true);
+                            break;
+                        case AR_AVAILABILITY_UNKNOWN_CHECKING:
+                            std::this_thread::sleep_for(std::chrono::milliseconds(200));
+                            break;
+                        default:
+                            tcs.complete(false);
+                            break;
+                    }
+
+                    if (tcs.completed())
+                    {
+                        break;
+                    }
+                }
+
+                if (!tcs.completed())
+                {
+                    tcs.complete(false);
+                }
+            });
+
+            return tcs.as_task();
+        }
+
+        // VR and inline sessions are not supported at this time.
+        return arcana::task_from_result<std::exception_ptr>(false);
+    }
+
+    bool CheckARCoreInstallStatus(bool requestInstall)
+    {
+        ArInstallStatus install_status;
+        ArStatus installStatus = ArCoreApk_requestInstall(
+                GetEnvForCurrentThread(), GetMainActivity(), requestInstall, &install_status);
+        return installStatus == AR_SUCCESS && install_status == AR_INSTALL_STATUS_INSTALLED;
+    }
+
+    arcana::task<void, std::exception_ptr> CheckAndInstallARCore(const Babylon::JsRuntimeScheduler& runtimeScheduler)
```

Comment:
I don't understand why all of these functions take a JsRuntimeScheduler. Do you actually need to be on the JsRuntime thread?

Reply:
The final constructor of session impl is the only one that must run in the context of the jsRuntime, as we must load shaders on the main JS thread.

I had these all using an inline_scheduler, but it was not running subsequent tasks when coming back from the nested tasks.  I figured since we are already passing down the jsRuntime I could just throw them on the main JsRuntime thread, as we already have a handle to that context.

Reply:
The final constructor of session impl is the only one that must run in the context of the jsRuntime, as we must load shaders on the main JS thread.

I had these all using an inline_scheduler, but it was not running subsequent tasks when coming back from the nested tasks. I figured since we are already passing down the jsRuntime I could just throw them on the main JsRuntime thread, as we already have a handle to that context.

Reply:
Oh, I just saw this comment, oops. I left a bunch of comments on this topic as well. Let's try `inline_scheduler` and if it isn't working for some reason I'm happy to take a look with you.

Reply:
Worked with Ryan offline, and got rid of the need to use another scheduler. There was a bug in our use of ticketed collection using a mutex instead of a recursive_mutex that was causing my other issues, and we moved the initialization code to the first instance of GetNextFrame to get around the OpenGL issues.